Main.hs

```diff
@@ -200,45 +183,48 @@ handleEvent (MessageUpdate Message{..}) = do
                                         queue <- use channel
                                         status <- liftIO $ writeUChan queue (messageChannel, messageId)
                                         when (status == Just False) $ do
-                                          sendTrace $ CreateReaction (messageChannel, messageId) react_wait
+                                          void $ restCallTrace $ CreateReaction (messageChannel, messageId) react_wait
     _ -> pure ()
 handleEvent (MessageDelete messageChannel messageId) = do
   pruneMessages
   x <- use (recentMsgs . at (messageChannel, messageId))
   case x of
-    Just (ts, oldreq, Just id) -> sendTrace $ DeleteMessage (messageChannel, id)
+    Just (ts, oldreq, Just id) -> void $ restCallTrace $ DeleteMessage (messageChannel, id)
     _ -> pure ()
 handleEvent _ = pure ()
 
 backendLoop :: EvalM ()
 backendLoop = forever $ catch (do queue <- use channel
                                   (chan, msg) <- liftIO $ readUChan queue
-                                  sendTrace $ DeleteOwnReaction (chan, msg) react_wait
+                                  void $ restCallTrace $ DeleteOwnReaction (chan, msg) react_wait
                                   x <- use (recentMsgs . at (chan, msg))
                                   case x of
                                     Just (_, req, reply) -> do
                                       outs <- forM req $ \cmd -> do
-                                        sendTrace $ TriggerTypingIndicator chan
+                                        void $ restCallTrace $ TriggerTypingIndicator chan
                                         case cmd of
                                           Reset mode -> do resetMode mode
                                                            pure ""
                                           EvalLine mode ln -> evalLine mode ln
                                           EvalBlock mode blk -> evalBlock mode blk
                                       let res = formatResults outs
                                       case reply of
-                                        Just id -> sendTrace $ EditMessage (chan, id) res Nothing
-                                        _ -> do Message{..} <- send $ CreateMessage chan res
-                                                assign (recentMsgs . at (chan, msg) . _Just . _3) (Just messageId)
+                                        Just id -> void $ restCallTrace $ EditMessage (chan, id) res Nothing
+                                        _ -> do res <- restCallTrace $ CreateMessage chan res
+                                                case res of
+                                                  Left e -> pure ()
+                                                  Right (Message{..}) ->
+                                                    assign (recentMsgs . at (chan, msg) . _Just . _3) (Just messageId)
```

Comment:
If the CreateMessage fails, we throw anyway, preventing the `assign`.

Reply:
What do you mean here? In the original `send $ CreateMessage chan res` has a `throwM` that prevents the `assign` on an error. In the pull request checking the result of `restCallTrace` is in-lined so `assign` only evaluates on a successful `Message`.